README.md
# kaph*

*kaph* is loose-coupled set of tools for handle requests under node.js.

*In the Phoenician alphabet letter "kaph" indicates palm.

## Design



ChangeLog
In work, Version 0.2

* Change API

* Docs

* Tests (in work)

* NPM package

Version 0.1

* Initial Version

http.js
var inherits = require('util').inherits;
var http = require('http');

/**
 * Handler for HTTP Request
 * @param {http.ServerRequest} request
 * @param {http.ServerResponse} response
 * @param {Array} chain chain of operations. See readme.
 * @param {Array} args Optional arguments
 * @returns {Handler}
 */
function Handler(request, response, chain, args) {
    this.request = request;
    this.response = response;
    this._args = args || [];
    
    this._chain = chain;
    this._level = -1; // Level in chain of operations
    
    this.logger = console; // default logger
};
exports.Handler = Handler;

/**
 * Executes next operation in chain. In operation selects the current 
 * request method ("GET", "POST" etc.) or "DEFAULT". 
 * 
 * If operation throws error - handles it. To generate a human readable error 
 * message uses method "ERROR" of operation.
 */
Handler.prototype.next = function() {
    this._level++;
    var op = this._chain[this._level];
    try {
        var meth = op[this.request.method] || op['DEFAULT'];
        if (!meth) throw new HandlerError(405, 'Operation hasn\'t methods ' + 
                this.request.method + ' or DEFAULT.');
        return meth.apply(this, this._args);
    } catch (e) {
        var errGen = op['ERROR'] || ERROR;
        this._handleError(e, errGen);
    }
};

Handler.prototype.error = function(code, reason) {
    var op = this._chain[this._level];
    var errGen = op['ERROR'] || ERROR;
    this._handleError(new HandlerError(code, reason), errGen);
};

/**
 * Handle error. 
 * @param e error
 * @param errGen Error message generator
 */
Handler.prototype._handleError = function(e, errGen) {
    var summary = this.request.method + " " + this.request.url + 
        ' (level ' + this._level + '):';
    var code = 0;
    var message = '';
    
    if (e instanceof HandlerError) {
        message = e.code in http.STATUS_CODES ? 
                summary + " " + e : 
                summary + "Bad HTTP status code " + e.code;
       
        code = e.code;
    } else {
        message = "Uncaught exception in " + summary + "\n" +
                (e.stack || e);
        code = 500;
    }
    this.logger.error(message);
    
    this.response.writeHead(code, 
            {'Content-Type': "text/html; charset=UTF-8"});
    this.response.end(errGen(code, message));
};

/**
 * Default function for generate error html body.
 * @param {Number} code 
 * @param {String} message
 * @returns {String}
 */
function ERROR(code, message) {
    return "<html><title>" + message + "</title>" +
            "<body><code>" + message + "</code></body></html>";
};
exports.ERROR = ERROR;

/**
 * HTTP Error
 * @constructor
 * @extends Error
 * @param code Optional status code. By default is "500"
 * @param reason Optional reason message
 */
function HandlerError(code, reason) {
    this.name = "HandlerError";
    this.code = code || 500;
    this.reason = reason || http.STATUS_CODES[this.code] || "Not implemented";
    this.message = this.code + ": " + this.reason;
    Error.captureStackTrace(this);
}
inherits(HandlerError, Error);
exports.HandlerError = HandlerError;

chain/cookie.js
var crypto = require('crypto');
var Buffer = require('buffer').Buffer;

exports.secret = hex_hmac_sha1(Math.random(), Math.random());

/**
 * Kaph stack Cookie operation. 
 */
Op = {
    DEFAULT: function() {
        this.cookies = new Proc(this.request.headers, this.response);
        return this.next();
    }
};
exports.Op = Op;

/**
 * Cookie processor.
 * @param {http.ServerRequest} request
 * @param {http.ServerResponse} response
 * @param {Boolean} dontDecorate Don't decorate 
 * @returns {Proc}
 */
function Proc(headers, response, dontDecorate) {
    this._headers = headers;
    this._response = response;
    this._dontDecorate = dontDecorate;
}
exports.Proc = Proc;

/**
 * Set cookie. If needed, at first call decorates method `writeHead` of 
 * instance `http.ServerResponse`.
 * @param {String} name 
 * @param value Cookie value
 * @param {Object} options Optional options. See readme.
 * @param {Boolean} encrypt Encrypt cookie value
 */
Proc.prototype.set = function(name, value, options, encrypt) {
    if (this._response._headerSent) return;
    options = options || {};
    if (!this._outgoing) {
        this._outgoing = {};
        
        if (!this._dontDecorate) {
            // Decorate original method
            var _writeHead = this._response.writeHead;
            var COOKIE_KEY = 'Set-Cookie', slice = Array.prototype.slice;
            var self = this;
            this._response.writeHead = function() {
                // Honor the passed args and method signature
                // (see http.writeHead docs)
                var args = slice.call(arguments), headers = args[args.length - 1];
                if (!headers || typeof (headers) != 'object') {
                    // No header arg - create and append to args list
                    args.push(headers = []);
                }
                
                // Merge cookie values
                var prev = headers[COOKIE_KEY], cookies = self.deploy() || [];
                if (prev) cookies.push(prev);
                if (cookies.length > 0)
                    headers[COOKIE_KEY] = cookies;
                
                // Invoke original writeHead()
                _writeHead.apply(this, args);
            };
        }
    }
    
    // determine expiration date
    if (options.days) {
        options.expires = options.expires || new Date();
        options.expires.setDate(options.expires.getDate() + options.days);
    }

    // Serve value
    value = (value !== null && typeof value !== 'undefined') ?
            value.toString() : '';
    
    if (encrypt) {
        // If value is secure - encrypt
        value = [value.length, encode(value), options.expires ];
        var signature = hex_hmac_sha1(value.join("|"), exports.secret);
        value.push(signature);
        value = value.join('|').replace(/=/g, '*');
    }

    // Form cookie
    var cookie = name + '=' + escape(value) + ';';
    options.expires && (cookie += ' expires=' + 
                options.expires.toUTCString() + ";");
    options.path && (cookie += ' path=' + options.path + ';');
    options.domain && (cookie += ' domain=' + options.domain + ';');
    options.secure && (cookie += ' secure;');
    options.httpOnly && (cookie += ' httponly');

    this._outgoing[name] = cookie;
};

/**
 * Get cookie by name
 * @param {String} name
 * @param {Boolean} decrypt
 */
Proc.prototype.get = function(name, decrypt) {
    // Parse cookies if not yet parsed
    if (!this._incoming) {
        var header = this._headers["cookie"] || "";
        var self = this;
        this._incoming = {};
        
        header.split(";").forEach( function( cookie ) {
            var parts = cookie.split("="),
            name =  (parts[0] ? parts[0].trim() : ''),
            value = (parts[1] ? parts[1].trim() : '');
            self._incoming[name] = unescape(value);
        });
    }
    
    var value = this._incoming[name];
    
    // Decript value if needed
    if (decrypt && value) {
        var parts = value.replace(/\*/g, '=').split("|");
        if (parts.length !== 4) {
            return;
        }
        
        var len = parts[0];
        value = decode(parts[1]).substr(0, len);
        var expires = new Date(+parts[2]);
        var remoteSig = parts[3];
        
        if ( expires < new Date ) {
            return;
        }
        
        var localSig = hex_hmac_sha1(parts.slice(0, 3).join("|"), 
                exports.secret);
        
        if ( localSig !== remoteSig ) {
            throw new Error("invalid cookie signature: " + name);
        }
        return value;
    }
    return value;
};

/**
 * Clears cookie
 * @param {String} name
 */
Proc.prototype.clear = function(name) {
    options = {expires: new Date( +new Date - 30 * 24 * 60 * 60 * 1000) };
    this.set(name, '', options);
};

/**
 * Generate "Set-Cookie" header value
 * @returns {Array}
 */
Proc.prototype.deploy = function() {
    if (!this._outgoing) return;
    var stream = [];
    for (var k in this._outgoing) {
        stream.push(this._outgoing[k]);
    }
    return stream;
};

/**
 * Generate hash.
 * @param data data
 * @returns {String} hash
 */
function hex_hmac_sha1(data, key) {
    var hmac = crypto.createHmac('sha1', key);
    hmac.update(data);
    return hmac.digest('hex');
}

/**
 * Encode data to base64
 * @param data
 * @returns {String}
 */
function encode(data) {
    return (new Buffer(data)).toString('base64');
}
/**
 * Decode data from base64
 * @param data
 * @returns {String}
 */
function decode(data) {
    return (new Buffer(data, 'base64')).toString('utf8');
}

chain/writer.js
var http = require('http');
var url = require('url');
var Buffer = require('buffer').Buffer;
var crypto = require('crypto');

/**
 * Kaph stack Writer operation. Just makes writer object in Handler.
 */
Op = {
    DEFAULT: function() {
        this.writer = new Writer(this.request, this.response);
        return this.next();
    }
};
exports.Op = Op;

/**
 * Writer facade.
 * @param {http.ServerRequest} request
 * @param {http.ServerResponse} response
 * @returns {Writer}
 */
function Writer(request, response) {
    this._request = request;
    this._response = response;
    
    this._encoding = 'utf8';
    this._statusCode = 200;
    this._headers = {'Content-Type': "text/html; charset=UTF-8"};
}
exports.Writer = Writer;

/**
 * Sets response status code. 
 * @param {Number} code HTTP Status code
 */
Writer.prototype.setStatus = function(code) {
    if (!this._response._headerSent && code in http.STATUS_CODES)
            this._statusCode = code;
};

/**
 * Set response header.
 * @param {String} name
 * @param value
 * @throws {Error}
 */
Writer.prototype.setHeader = function(name, value) {
    if (this._response._headerSent) return;
    value = value ? value.toString() : '';
    if (value != value.replace(/[\x00-\x1f]/, " ").substring(0, 4000))
        throw new Error('Unsafe header value ' + value);
    this._headers[name] = value;
};

/**
 * Write to response
 * @param data
 * @param encoding
 */
Writer.prototype.write = function(data, encoding) {
    if (!data) return; // If no data - do nothing
    if (!Buffer.isBuffer(data) && typeof data !== 'string')
            data = data.toString();
    if (!this._response._headerSent) this._sendHeaders();
    this._response.write(data, (encoding || this._encoding));
};

/**
 * 
 * @param data
 * @param encoding
 */
Writer.prototype.end = function(data, encoding) {
    if (data && !Buffer.isBuffer(data) && typeof data !== 'string')
            data = data.toString();
    if (!this._response._headerSent) {
        // if any data present - add some info in headers:
        if (!!data && this._statusCode == 200 
                && this._request.method == 'GET') {
            // ETag
            if (!('ETag' in this._headers)) {
                etag = '"' + crypto.createHash("sha1").
                        update(data).digest("hex") + '"';
                
                // Check if-none-match
                var inm = this._request.headers["if-none-match"];
                if (inm && inm.indexOf(etag) != -1) {
                    // Not modified - just send 304
                    this._response.writeHead(304);
                    this._response.end();
                    return;
                } else {
                    this.setHeader("ETag", etag);
                }
            }
            
            // Content-Length
            if (!("Content-Length" in this._headers)) {
                var l = Buffer.isBuffer(data) ? data.length 
                        : Buffer.byteLength(data, encoding || this._encoding);
                this.setHeader("Content-Length", l);
            }
        }
        this._sendHeaders();
    }
    this._response.end(data, encoding || this._encoding);
};

/**
 * Sends a redirect to the given (optionally relative) URL.
 * @param redirectUrl Redirect URL
 * @param permanent Permanent. Default = false
 * @throws HttpError
 */
Writer.prototype.redirect = function(redirectUrl, permanent) {
    permanent = permanent || false;
    if (this._response._headerSent) {
        throw new Error('Cannot redirect after headers have been written');
    }
    this.setStatus(permanent ? 301 : 302);
    redirectUrl = redirectUrl.replace(/[\x00-\x1f]/, "");
    this.setHeader('Location', url.resolve(this._request.url, redirectUrl));
    this.end();
};

/**
 * Sends all headers to client.
 */
Writer.prototype._sendHeaders = function() {
    this._response.writeHead(this._statusCode, this._headers);
};

index.js
/**
 * Kaph. Routerless HTTP handler for node.js
 */

var inherits = require('util').inherits;
var http = require('http');
var url = require('url');
var Buffer = require('buffer').Buffer;
var crypto = require('crypto');

/**
 * Kaph request handler.
 * 
 * @constructor
 * 
 * @param request Original request
 * @param response Original clear response
 * @param args Optional Array of arguments
 * @type args Array
 * @returns {Handler}
 */
function Handler(request, response, args) {
    // request
    this.request = request;
    //response
    this.response = response;
    // arguments
    this.args = args || [];
    
    this._headersWritten = false; // Headers not written
    this.clear(); // clear all
}
exports.Handler = Handler;

/**
 * Cleans response.
 */
Handler.prototype.clear = function() {
    this._headers = {
            "Server": "node.js:" + process.version,
            "Content-Type": "text/html; charset=UTF-8"
    };
    this._statusCode = 200;
    this._encoding = 'utf8';
};

// Supported methods
Handler.prototype._supportedMethods = 
        ['HEAD', 'GET', 'POST', 'PUT', 'DELETE', 'OPTIONS'];

/**
 * HEAD method. Must be implemented.
 * @throws HttpError
 */
Handler.prototype.HEAD = function() {
    throw new HttpError(405, "HEAD not implemented");
};

/**
 * GET method. Must be implemented.
 * @throws HttpError
 */
Handler.prototype.GET = function() {
    throw new HttpError(405, "GET not implemented");
};

/**
 * POST method. Must be implemented.
 * @throws HttpError
 */
Handler.prototype.POST = function() {
    throw new HttpError(405, "POST not implemented");
};

/**
 * PUT method. Must be implemented.
 * @throws HttpError
 */
Handler.prototype.PUT = function() {
    throw new HttpError(405, "PUT not implemented");
};

/**
 * DELETE method. Must be implemented.
 * @throws HttpError
 */
Handler.prototype.DELETE = function() {
    throw new HttpError(405, "DELETE not implemented");
};

/**
 * OPTIONS method. Must be implemented.
 * @throws HttpError
 * 
 */
Handler.prototype.OPTIONS = function() {
    throw new HttpError(405, "OPTIONS not implemented");
};

/**
 * Generate error HTML page.
 * @param code Status code
 * @param message Message
 * @returns {String} Error html
 */
Handler.prototype.getErrorHtml = function(code, message) {
    var msg = code + ": " + message; 
    return "<html><title>" + msg + "</title>" +
            "<body>" + msg + "</body></html>";
};

/**
 * Sets response status code.
 * @param code Status code
 */
Handler.prototype.setStatus = function(code) {
    if (code in http.STATUS_CODES) this._statusCode = code;
};

/**
 * Sets the given response header name and value.
 * @param name Header name
 * @param value Header value
 * @throws HttpError
 */
Handler.prototype.setHeader = function(name, value) {
    var value = value.toString();
    var safe = value.replace(/[\x00-\x1f]/, " ").substring(0, 4000);
    if (safe != value) throw new Error('Unsafe header value ' + value);
    this._headers[name] = value;
};

/**
 * Sends all headers to client.
 */
Handler.prototype.sendHeaders = function() {
    this._headersWritten = true;
    this.response.writeHead(this._statusCode, this._headers);
};

/**
 * Writes data to client. 
 * If headers not sent - sends they first.
 * @param {Buffer||String} Data chunk. If no data - nothing happens. 
 * @param {String} encoding
 */
Handler.prototype.write = function(data, encoding) {
    if (!data || data == null) return; // If no data - do nothing
    
    if (!this._headersWritten) {
        this.sendHeaders();
    }
//    if (data instanceof Object) {
//        this.setHeader("Content-Type", "text/javascript; charset=UTF-8");
//        data = JSON.stringify(data);
//    }
        
    this.response.write(data, (encoding || this._encoding));
};

/**
 * Ends response.
 * @param data
 * @param encoding
 */
Handler.prototype.end = function(data, encoding) {
    data = data || "";
    if (!this._headersWritten) {
        // If headers not written yet - add some info in headers:
        
        // ETags
        if (this._statusCode == 200 && this.request.method == 'GET' &&
                !('ETag' in this._headers) && have_openssl) {
            etag = '"' + crypto.createHash("sha1").
                    update(data).digest("hex") + '"';
            
            var inm = this.request.headers["if-none-match"];
            if (inm && inm.indexOf(etag) != -1) {
                // Not modified - just send 304
                this.response.writeHead(304);
                this.response.end("");
                return;
            } else {
                this.setHeader("ETag", etag);
            }
            
            // and content length
            if (!("Content-Length" in this._headers)) {
                var l = Buffer.isBuffer(data) ? data.length 
                        : Buffer.byteLength(data, encoding || this._encoding);
                this.setHeader("Content-Length", l);
            }
        }
        this.sendHeaders();
    }    
    this.response.end(data, encoding || this._encoding);
};

/**
 * Sends a redirect to the given (optionally relative) URL.
 * @param redirectUrl Redirect URL
 * @param permanent Permanent. Default = false
 * @throws HttpError
 */
Handler.prototype.redirect = function(redirectUrl, permanent) {
    permanent = permanent || false;
    if (this._headersWritten) {
        throw new Error('Cannot redirect after headers have been written');
    }
    this.setStatus(permanent ? 301 : 302);
    redirectUrl = redirectUrl.replace(/[\x00-\x1f]/, "");
    this.setHeader('Location', url.resolve(this.request.url, redirectUrl));
    this.end();
};

/**
 * Sends the given HTTP error code to the browser.
 * @param code Error code
 * @param message Error Message
 */
Handler.prototype.sendError = function(code, message) {
    this.clear();
    this.setStatus(code);
    this.end(this.getErrorHtml(code, message));
};

/**
 * Executes handler. Catches all errors
 */
Handler.prototype.execute = function() {
    try {
        if (this._supportedMethods.indexOf(this.request.method) == -1)
            throw new HttpError(405);
        
        this[this.request.method].apply(this, this.args);
    } catch (e) {
        this.handleError(e);
    }
};

/**
 * Handles all exceptions in request. 
 * Logs it and send to client if possible.
 * @param e Exception
 */
Handler.prototype.handleError = function(e) {
    if (e instanceof HttpError) {
        if (e.code in http.STATUS_CODES) {
            console.error(this._summary() + " " + e);
            this.sendError(e.code, e.reason);
        } else {
            console.error("Bad HTTP status code %d", e.code);
            this.sendError(e.code, e.reason);
        }
    } else {
        console.error("Uncaught exception in " + this._summary() + "\n" +
                (e.stack || e));
        this.sendError(500, (e.stack || e));
    }
};

/**
 * Get request summary.
 * @returns {String}
 */
Handler.prototype._summary = function() {
    return this.request.method + " " + this.request.url;
};

/**
 * HTTP Error
 * @constructor
 * @extends Error
 * @param code Status code
 * @param reason Reason
 */
function HttpError(code, reason) {
    //Error.call(this);
    this.name = "HttpError";
    this.code = code || 500;
    this.reason = reason || http.STATUS_CODES[this.code] || "Not implemented";
    this.message = this.code + " " + this.reason;
    Error.captureStackTrace(this);
}
inherits(HttpError, Error);
exports.HttpError = HttpError;
